src/java/main/org/apache/zookeeper/common/TokenBucket.java

```diff
@@ -0,0 +1,88 @@
+/**
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package org.apache.zookeeper.common;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Simple rate limiter based on a token bucket.
+  */
+public class TokenBucket implements RateLimiter {
+
+    // VisibleForTesting
+    long refreshPeriodNanos = TimeUnit.SECONDS.toNanos(1L);
+    // VisibleForTesting
+    volatile long nextRefillTime;
+    private AtomicLong tokens;
+    private long capacity;
+    private long tokensPerPeriod;
+
+    public TokenBucket() { }
+
+    // VisibleForTesting
+    public TokenBucket(long capacity, long tokensPerSecond, long initialTokens) {
+        this.tokens = new AtomicLong(initialTokens);
+        this.capacity = Math.max(capacity, tokensPerSecond);
+        this.tokensPerPeriod = tokensPerSecond;
+        this.nextRefillTime = System.nanoTime() + refreshPeriodNanos;
```

Comment:
nit: do we need nano seconds? can we just use org.apache.zookeeper.common.Time.currentElapsedTime() ?